### PR TITLE
Add support for Random Serial Numbers v3

### PR DESCRIPTION
**Pull request by rcritten**
This is a temporary PR to run the rawhide CI with RSN enabled. I'm trying to shake out any bugs that didn't appear in my local testing.

### CODE REVIEW

**Comment by rcritten** [comment]
The Azure failures are expected because the required PKI packages are only in rawhide.

**Comment by rcritten** [comment]
The failures so far break down into a few categories:
* expecting a specific serial number, both in comparing return values and ipa cert-show 1
* some issue in how errors are reported/expected in the UI
* ipa-ca-install doesn't work in CI though it worked for me in manual testing
* other non-related known issues (DNS, fips mode not working, an smb error)

**Comment by rcritten** [comment]
rerun: vagrant provisioning failed

**Comment by rcritten** [comment]
rerun: vagrant provisioning failed

**Comment by rcritten** [comment]
Provisioning is failing because of repo issues:

Depsolve Error occured:
 Problem 1: cannot install the best candidate for the job
  - nothing provides samba-client-libs >= 2:4.16.1-1.fc37 needed by freeipa-server-4.10.0.dev-0.fc37.x86_64
 Problem 2: package freeipa-server-trust-ad-4.10.0.dev-0.fc37.x86_64 requires freeipa-server = 4.10.0.dev-0.fc37, but none of the providers can be installed
  - cannot install the best candidate for the job
  - nothing provides samba-client-libs >= 2:4.16.1-1.fc37 needed by freeipa-server-4.10.0.dev-0.fc37.x86_64
 Problem 3: package freeipa-server-dns-4.10.0.dev-0.fc37.noarch requires freeipa-server = 4.10.0.dev-0.fc37, but none of the providers can be installed
  - cannot install the best candidate for the job
  - nothing provides samba-client-libs >= 2:4.16.1-1.fc37 needed by freeipa-server-4.10.0.dev-0.fc37.x86_64
 Problem 4: package freeipa-server-4.9.9-1.fc37.x86_64 requires freeipa-client = 4.9.9-1.fc37, but none of the providers can be installed
  - package freeipa-client-4.10.0.dev-0.fc37.x86_64 conflicts with ipa-admintools provided by freeipa-client-4.9.9-1.fc37.x86_64
  - package freeipa-client-4.10.0.dev-0.fc37.x86_64 conflicts with ipa-client provided by freeipa-client-4.9.9-1.fc37.x86_64
  - package freeipa-client-4.9.9-1.fc37.x86_64 conflicts with ipa-admintools provided by freeipa-client-4.10.0.dev-0.fc37.x86_64
  - package freeipa-client-4.9.9-1.fc37.x86_64 conflicts with ipa-client provided by freeipa-client-4.10.0.dev-0.fc37.x86_64
  - cannot install both freeipa-client-4.9.9-1.fc37.x86_64 and freeipa-client-4.10.0.dev-0.fc37.x86_64
  - package freeipa-healthcheck-0.10-1.fc36.noarch requires freeipa-server, but none of the providers can be installed
  - cannot install the best candidate for the job
  - conflicting requests
  - nothing provides samba-client-libs >= 2:4.16.1-1.fc37 needed by freeipa-server-4.10.0.dev-0.fc37.x86_64

**Comment by abbra** [comment]
I think this run came around the time rawhide pushes were done. I restarted failed runs in PRCI and some of them now fail in actual tests, installation phase works just fine.

**Comment by rcritten** [comment]
test_cert_fix is failing due to https://github.com/dogtagpki/pki/issues/3996

**Comment by rcritten** [comment]
I have a fix in hand for the Azure XML-RPC failure. It was related to strings I changed to fix a different test.

**Comment by rcritten** [comment]
I fixed another XML-RPC failure, I think the last, due to missing ipacarandomserialnumberversion in the CA tracker.

The candidate integration test has been running for 3 hours so maybe I need to pare it down a bit.

**Comment by rcritten** [comment]
test_integration/test_commands.py::TestIPACommand::test_ssh_key_connection failed and I can't see how it is related.
test_integration/test_commands.py::TestIPACommand::test_ssh_leak because the previous test failed.

I think the last thing to do before resubmitting as a new, clean PR is to skip if the PKI version is < 11.2.0. The PKI am tells me there is no property to be used to determine the capability.

**Comment by rcritten** [comment]
I squashed all the patches so things are a bit cleaner.

Left to do is to determine how to test this in the long term. We can extend the rawhide_nightly tests with these (preferred by me) or as a separate execution (or something I haven't considered).

I have not yet committed to the new LDAP attribute to store the RSN version number because I wanted it to go through review first. This is necessary because PKI does not provide a mechanism for querying whether RSN is enabled or not. It is only stored in config files.

Replacing the method I came up with another one should be relatively straightforward so if you hate it let me know.

File: install/share/60certificate-profiles.ldif [hunk 1]
```diff
@@ -7,6 +7,7 @@ attributeTypes: (2.16.840.1.113730.3.8.21.1.5 NAME 'ipaCertProfileCategory' DESC
 attributeTypes: (2.16.840.1.113730.3.8.21.1.6 NAME 'ipaCaId' DESC 'Dogtag Authority ID' EQUALITY caseIgnoreMatch ORDERING caseIgnoreOrderingMatch SUBSTR caseIgnoreSubstringsMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.15 X-ORIGIN 'IPA v4.4 Lightweight CAs' )
 attributeTypes: (2.16.840.1.113730.3.8.21.1.7 NAME 'ipaCaIssuerDN' DESC 'Issuer DN' SUP distinguishedName X-ORIGIN 'IPA v4.4 Lightweight CAs' )
 attributeTypes: (2.16.840.1.113730.3.8.21.1.8 NAME 'ipaCaSubjectDN' DESC 'Subject DN' SUP distinguishedName X-ORIGIN 'IPA v4.4 Lightweight CAs' )
+attributeTypes: (2.16.840.1.113730.3.8.21.1.9 NAME 'ipaCaRandomSerialNumberVersion' DESC 'Random Serial Number Version' EQUALITY integerMatch ORDERING integerOrdingMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 SINGLE-VALUE X-ORIGIN 'IPA v4.9 RSNv3' )
```
Comment (f-trivino) [inline install/share/60certificate-profiles.ldif:10]: uhmm... integerOrdingMatch ---> integerOrderingMatch ??

ERR - attr_syntax_create - Error: the ORDERING matching rule [integerOrdingMatch] is not compatible with the syntax [1.3.6.1.4.1.1466.115.121.1.27] for the attribute [ipaCaRandomSerialNumberVersion]

**Comment by abbra** [comment]
Use of an LDAP attribute is OK. It still doesn't prevent a CA replica that is built using older PKI version -- I guess PKI would break itself in that case and deployment would fail?

API version bump is missing and please make sure to add the new attribute to the OIDs database.

Also, please add a design document that explains logic and schema use.

**Comment by rcritten** [comment]
If you mean that there is nothing to prevent a user from adding the LDAP attribute themselves to work around the check then yes, there is no way to prevent it. But from a pure IPA installation tools no interoperability between ranged and RSN servers is allowed.

**Comment by abbra** [comment]
LGTM. Please remove the temp commit.

**Comment by rcritten** [comment]
@abbra there is still the question of where to put the subclassed tests. Into rawhide/nightly, normal nightly, both?

**Comment by abbra** [comment]
I think both normal and rawhide nightlies.

**Comment by rcritten** [comment]
Nothing is written to disk in an RSN server, at least on the IPA side, that isn't written in a non-RSN server. Arguably PKI could be writing something so...maybe? I'm worried about extending the test times too far.

As to the wrong test names (lacking _RSN) I"m going to create a new temp commit to ensure they actually pass since the wrong ones were running.

**Comment by rcritten** [comment]
Temp commit tests are passing. Removing temp commit.

- fedora-rawhide/test_caless_TestServerCALessToExternalCA_RSN  http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/95d8aede-e2ba-11ec-8abd-fa163e33b8c8
- fedora-rawhide/test_random_serial_numbers_TestInstallWithCA_DNS1_RSN http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/a61680f8-e2bc-11ec-a925-fa163ec85172
- fedora-rawhide/test_random_serial_numbers_TestInstallWithCA_KRA1_RSN http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/b877b406-e2bc-11ec-93db-fa163ea4a9f1 
- fedora-rawhide/test_random_serial_numbers_test_commands_RSN http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/a92dd124-e2bc-11ec-83d9-fa163e172501

**Comment by rcritten** [comment]
I broke the pki version validator addressing a test failure. Investigating.

**Comment by rcritten** [comment]
I believe I've addressed the version checking failures you pointed out.

**Review by flo-renaud — COMMENTED** [review]
Hi @rcritten 
thanks for the update. Most of the issues were addressed but I spotted a (last?) one, please see my inline comment.
Apart from that, nightly_latest_testing.yaml and nightly_previous.yaml do not launch the new tests, is it on purpose?

File: ipaserver/install/ca.py [hunk 2]
```diff
@@ -165,6 +219,18 @@ def install_check(standalone, replica_config, options):
         options._subject_base = str(replica_config.subject_base)
         options._ca_subject = lookup_ca_subject(_api, options._subject_base)
 
+        options._random_serial_numbers = (
```
[the remark this answers is not in the document]
Comment (rcritten) [reply ipaserver/install/ca.py:222]: Fixed

**Comment by rcritten** [comment]
The other missing nightly tests were an oversight. I added both of them.

**Comment by flo-renaud** [comment]
@rcritten 
thanks for the update, LGTM. The installation of a replica without CA does not check any more for the support of RSNv3.

**Comment by rcritten** [comment]
Dropped the temp commit.

**Comment by flo-renaud** [comment]
master:

* d3481449ee77c4694bfc25dfa194d5875ffb4f74 doc/designs: add Random Serial Numbers v3 support
* 83be923ac566df53472e2b7b15814f1a1d00e933 Add a new parameter type, SerialNumber, as a subclass of Str
* beaa0562dc0117a4a8dfc96670b72af319d2cfab Add support for Random Serial Numbers v3
* d241d7405f49ba56eaa86acb052b38d76b23f697 Add tests for Random Serial Number v3 support

**Comment by flo-renaud** [comment]
@rcritten the PR needs a manual backport to ipa-4-9, likely because of the yaml test definitions. Can you do it? Thanks

**Comment by abbra** [comment]
@flo-renaud actually, we should not push this to ipa-4-9. We are going to branch off to ipa-4-10 and have RSN only in 4.10 series. I am going to write about it soon (we discussed it on Monday).

**Comment by flo-renaud** [comment]
@abbra thanks for the clarification, and sorry for the noise. So we're all done for this PR.

**Pull request closed without merging**